### DISCUSSION TITLE
Report deprecated dependencies

### DISCUSSION THREAD

**Discussion by jfmengels** [jfmengels/elm-review#92] [Rule ideas]
**What the rule should do:**

Report deprecated dependencies. I see a few packages in the registry with `DEPRECATED` in the `elm.json` summary. I am thinking that we could use that as a sign that package is deprecated, and that it shouldn't be used anymore. Maybe we will be able to find other clues, like titles in the README containing `deprecated`.

**What problems does it solve:**

Allowing you to detect when a package becomes deprecated, in order to use the latest/best package out there for a problem.

**Example of things the rule would report:**

- A dependency of the project's summary is "DEPRECATED: Use xyz instead"

**When (not) to enable this rule:**

**I am looking for:**

Answers to the following questions:
- Should this rule take as a configuration a list of dependencies to allow? I'm thinking no, but if/when the package becomes deprecated, you may not want to fix it right now, as it can be a tedious/long task, depending on the package, meaning that you'd have to ignore the rule for a while. If we allow configuration, that will allow the users to keep the rule enabled and notice new deprecated dependencies.

- What do you think of the proposed deprecated signs? Can you think of others?

I am thinking it would fit very well with https://github.com/jfmengels/elm-review-rule-ideas/issues/5 (which we have already enabled at my workplace), in something like `elm-review-dependencies`. But we might want to have a similar rule for noticing deprecated modules/functions, in which case they wouldn't be in the same package. And maybe that's okay.

**Comment by lue-bird** [comment]
I've seen/used
- archived github repo
- "deprecated" or "moved to/into ..." or "republished/continued under ..." or "unmaintained" or "usage discouraged" or "use/using ... instead" in the first non-title line / title / package description
- (maybe when the package only contains 1 [dummy module](https://package.elm-lang.org/packages/indique/elm-n-nat/latest/Moved))

There are some signs where a program can be sure that the package is deprecated.
It's not so clear for others. Here's how I'd do it:
- write a tool (an elm-review rule?) that lists all potential deprecated packages
  - adds packages that are deprecated for sure to the `deprecatedPackages` set and
  - asks you to manually check packages where it isn't 100% sure:
    - you're sure → adds packages to the `deprecatedPackages`
- automatically run that tool after some time again for the updated/newly released packages
```elm
deprecatedPackages =
    [ "jfmengels/review-common"
    , ...
    ]
        |> Set.fromList

isDeprecated packageName =
    Set.member packageName deprecatedPackages
```
